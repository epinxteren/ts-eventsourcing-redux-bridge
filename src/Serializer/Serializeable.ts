import 'reflect-metadata';
import { Metadata } from 'ts-eventsourcing/Metadata';
import { ClassConstructor } from './ClassConstructor';

const SERIALIZER = Symbol.for('serializer');

export interface SerializeMetadata {
  normalize?: string;
  deNormalize?: string;
}

export function getSerializerMetadata(constructor: ClassConstructor): SerializeMetadata {
  const metadata: SerializeMetadata = Metadata.getMetadata(SERIALIZER, constructor);
  if (metadata) {
    return metadata;
  }
  const parent = Object.getPrototypeOf(constructor.prototype).constructor;
  if (parent === Object) {
    return {};
  }
  return Object.assign({}, getSerializerMetadata(parent));
}

/**
 * Decorator function for object to normalize it.
 */
export function Normalize(target: { constructor: ClassConstructor } | any, functionName: string): void {
  const constructor = target.constructor;
  const metadata: SerializeMetadata = getSerializerMetadata(constructor);
  metadata.normalize = functionName;
  Metadata.defineMetadata(SERIALIZER, metadata, constructor);
}

/**
 * Decorator static function to de-normalize it.
 */
export function DeNormalize(constructor: ClassConstructor, functionName: string): void {
  const metadata: SerializeMetadata = getSerializerMetadata(constructor);
  metadata.deNormalize = functionName;
  Metadata.defineMetadata(SERIALIZER, metadata, constructor);
}
