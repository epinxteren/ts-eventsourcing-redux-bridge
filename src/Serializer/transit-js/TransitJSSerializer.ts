/// <reference path="./transit-immutable-js.d.ts" />

import * as transit from 'transit-immutable-js';
import { Record } from 'immutable';
import { SerializerInterface } from '../SerializerInterface';

export interface RecordConstructor {
  new(...args: any[]): Record<any>;
}

export class TransitJSSerializer implements SerializerInterface {

  private readonly recordTransit: any;

  constructor(records: RecordConstructor[], extraHandlers: any[] = []) {
    /**
     * Simple map to verify you cannot give a record with the same name.
     */
    const recordsTypes: { [key: string]: RecordConstructor } = {};
    records.forEach(record => {
      const descriptiveName = Record.getDescriptiveName(new record());
      if (descriptiveName === '' || descriptiveName.toLowerCase() === 'record') {
        throw new Error(`wrong descriptiveName record name ${descriptiveName}`);
      }
      if (typeof recordsTypes[descriptiveName] !== 'undefined') {
        throw new Error(`Records with this name already given ${descriptiveName}`);
      }
      recordsTypes[descriptiveName] = record;
    });
    const withRecords = transit.withRecords(records);
    this.recordTransit = withRecords.withExtraHandlers(extraHandlers);
  }

  public serialize(data: unknown): string {
    return this.recordTransit.toJSON(data);
  }

  public deserialize(json: string): unknown {
    return this.recordTransit.fromJSON(json);
  }

}
