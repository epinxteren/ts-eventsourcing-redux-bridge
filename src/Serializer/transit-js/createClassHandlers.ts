import { ClassConstructor } from './TransitJSSerializer';

interface SerializeHandler {
  /**
   * a unique identifier for this type that will be used in the serialised output
   */
  tag: string;
  /**
   * a constructor function that can be used to identify the type via an instanceof check
   */
  class: ClassConstructor;
  /**
   * a function which will receive an instance of your type, and is expected to create some serialisable representation of it.
   */
  write(instance: any): any;
  /**
   * a function which will receive the serialisable representation, and is expected to create a new instance from it
   */
  read(instance: any): any;
}

export function createClassHandlers(classes: { [key: string]: ClassConstructor } = {}): SerializeHandler[] {
  return Object.getOwnPropertyNames(classes).map(tag => {
    const cons = classes[tag];
    return {
      tag,
      class: cons,
      write: (object: { [key: string]: any }) => {
        const ownPropertyNames = Object.getOwnPropertyNames(object);
        const plainProperties: { [key: string]: any } = {};
        ownPropertyNames.forEach(key => {
          plainProperties[key] = object[key];
        });
        return plainProperties;
      },
      read: (plainProperties: { [key: string]: any }) => {
        const instance = new cons();
        const ownPropertyNames = Object.getOwnPropertyNames(plainProperties);
        ownPropertyNames.forEach(key => {
          instance[key] = plainProperties[key];
        });
        return instance;
      },
    };
  });
}
