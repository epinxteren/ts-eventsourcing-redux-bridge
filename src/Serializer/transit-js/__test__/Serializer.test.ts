/* tslint:disable:max-classes-per-file */

import 'jest';
import { TransitJSSerializer } from '../TransitJSSerializer';
import { Record } from 'immutable';
import { createClassHandlers } from '../createClassHandlers';
import { DeNormalize, Normalize } from '../../Serializeable';

describe('Serializer', () => {

  it('Can serialize normal javascript objects', () => {
    const serializer = new TransitJSSerializer([]);
    const date = new Date();
    date.setTime(1518770045540);
    const target = {
      date,
      arrays: [1, 2, 3],
      objects: { test: 2 },
    };
    const serialized = serializer.serialize(target);
    expect(serialized).toMatchSnapshot();
    const deSerialized = serializer.deserialize(serialized);
    expect(deSerialized).toEqual(target);
  });

  it('Can serialize records', () => {
    const testRecord = Record({ name: 'foo' }, 'test');
    const test = new testRecord({ name: 'bar' });
    const serializer = new TransitJSSerializer([testRecord]);
    const serialized = serializer.serialize(test);
    expect(serialized).toMatchSnapshot();
    const deSerialized = serializer.deserialize(serialized);
    expect(deSerialized).toEqual(test);
  });

  it('Can serialize a random class', () => {
    class TestClass {
      public t = 2;
      public name = 55;
      public appel: string | undefined;
      public date: Date | undefined;

      constructor(public foo: string = 'bar') {

      }
    }
    const test = new TestClass('d');
    const date = new Date();
    date.setTime(1518770045540);
    test.date = date;
    const serializer = new TransitJSSerializer([], createClassHandlers({ TestClass }));
    const serialized = serializer.serialize(test);
    expect(serialized).toMatchSnapshot();
    const deSerialized: TestClass = serializer.deserialize(serialized) as any;
    expect(deSerialized).toEqual(test);
    expect(deSerialized).toBeInstanceOf(TestClass);
    expect(deSerialized.date).toBeInstanceOf(Date);
  });

  it('Can serialize a different classes', () => {
    class TestBase {
      public t = 2;
      public name = 55;
      public appel: string | undefined;
      public date: Date | undefined;

      constructor(public foo: string = 'bar') {

      }
    }

    class TestChild extends TestBase {

    }
    const test = new TestBase('d');
    const test2 = new TestChild('e');
    const date = new Date();
    date.setTime(1518770045540);
    test.date = date;
    const serializer = new TransitJSSerializer([], createClassHandlers({ TestClass2: TestChild, TestClass: TestBase }));
    const serialized = serializer.serialize([test, test2]);
    const deSerialized = serializer.deserialize(serialized) as any;
    expect(deSerialized).toEqual([test, test2]);
    expect(deSerialized[0]).toBeInstanceOf(TestBase);
    expect(deSerialized[1]).toBeInstanceOf(TestChild);
    expect(deSerialized[0]).not.toBeInstanceOf(TestChild);
  });

  class TestObject {

    @DeNormalize
    protected static deNormalize(data: any) {
      return new this(data.v);
    }
    constructor(public value: number) {
    }

    @Normalize
    protected normalize(): any {
      return { v: this.value };
    }
  }

  it('Can serialize with custom serializer functions', () => {
    const serializer = new TransitJSSerializer([], createClassHandlers({ TestObject }));
    const serialized = serializer.serialize(new TestObject(2323));

    expect(serialized).toMatch('["~#TestObject",["^ ","v",2323]]');

    const deSerialized = serializer.deserialize(serialized);
    expect(deSerialized).toBeInstanceOf(TestObject);
    expect(deSerialized).toEqual(new TestObject(2323));

  });

  it('Can serialize with custom serializer functions of extended class', () => {

    class TestObject2 extends TestObject {
    }

    const serializer = new TransitJSSerializer([], createClassHandlers({ TestObject2 }));
    const serialized = serializer.serialize(new TestObject2(2323));

    expect(serialized).toMatch('["~#TestObject2",["^ ","v",2323]]');

    const deSerialized = serializer.deserialize(serialized);
    expect(deSerialized).toBeInstanceOf(TestObject2);
    expect(deSerialized).toEqual(new TestObject2(2323));

  });

  it('Can serialize nested classes', () => {

    class Cat {
      constructor(public legs: Leg[]) {
      }

    }

    class Leg {
      constructor(public claws: Claw, public hairColor: string) {
      }
    }

    class Claw {

    }

    const serializer = new TransitJSSerializer([], createClassHandlers({ Cat, Leg, Claw }));
    const serialized = serializer.serialize(new Cat([
      new Leg(new Claw(), 'red'),
      new Leg(new Claw(), 'brown'),
      new Leg(new Claw(), 'white'),
      new Leg(new Claw(), 'black'),
    ]));

    const deserialized = serializer.deserialize(serialized);

    expect(deserialized).toEqual(new Cat([
      new Leg(new Claw(), 'red'),
      new Leg(new Claw(), 'brown'),
      new Leg(new Claw(), 'white'),
      new Leg(new Claw(), 'black'),
    ]));

  });

});
