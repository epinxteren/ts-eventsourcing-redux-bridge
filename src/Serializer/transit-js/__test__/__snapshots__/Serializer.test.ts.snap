// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serializer Can serialize a different classes 1`] = `"[[\\"~#TestClass\\",[\\"^ \\",\\"foo\\",\\"d\\",\\"t\\",2,\\"name\\",55,\\"date\\",\\"~m1518770045540\\"]],[\\"^0\\",[\\"^ \\",\\"foo\\",\\"e\\",\\"t\\",2,\\"^1\\",55]]]"`;

exports[`Serializer Can serialize a random class 1`] = `"[\\"~#TestClass\\",[\\"^ \\",\\"foo\\",\\"d\\",\\"t\\",2,\\"name\\",55,\\"date\\",\\"~m1518770045540\\"]]"`;

exports[`Serializer Can serialize normal javascript objects 1`] = `"[\\"^ \\",\\"date\\",\\"~m1518770045540\\",\\"arrays\\",[1,2,3],\\"objects\\",[\\"^ \\",\\"test\\",2]]"`;

exports[`Serializer Can serialize records 1`] = `"[\\"~#iR\\",[\\"^ \\",\\"n\\",\\"test\\",\\"v\\",[\\"^ \\",\\"name\\",\\"bar\\"]]]"`;
